File: onDeck/Models/PlayerState.swift
import Foundation

enum PlayerState {
    case active(GameContext)
    case upcoming(startTime: Date)
    case inactive(reason: InactiveReason)

    enum ActiveRole {
        case batting
        case pitching
    }

    struct GameContext {
        let gamePk: Int
        let role: ActiveRole
        let inning: String
        let homeTeam: String
        let awayTeam: String
        let homeTeamID: Int
        let awayTeamID: Int
        let homeScore: Int
        let awayScore: Int
        let balls: Int
        let strikes: Int
        let outs: Int
        let runnerOnFirst: Bool
        let runnerOnSecond: Bool
        let runnerOnThird: Bool
    }

    enum InactiveReason {
        case gameOver(gamePk: Int)
        case dayOff
        case substituted(gamePk: Int)
    }
}

File: windows/src/OnDeck.Core/Models/PlayerState.cs
namespace OnDeck.Core.Models;

/// <summary>
/// Port of <c>Models/PlayerState.swift</c>. Swift's enum-with-associated-values becomes a
/// closed record hierarchy: the private constructor keeps the case list to the nested types.
/// </summary>
public abstract record PlayerState
{
    private PlayerState() { }

    public sealed record Active(GameContext Context) : PlayerState;

    public sealed record Upcoming(DateTimeOffset StartTime) : PlayerState;

    public sealed record Inactive(InactiveReason Reason) : PlayerState;

    public enum ActiveRole
    {
        Batting,
        Pitching,
    }

    public sealed record GameContext(
        int GamePk,
        ActiveRole Role,
        string Inning,
        string HomeTeam,
        string AwayTeam,
        int HomeTeamId,
        int AwayTeamId,
        int HomeScore,
        int AwayScore,
        int Balls,
        int Strikes,
        int Outs,
        bool RunnerOnFirst,
        bool RunnerOnSecond,
        bool RunnerOnThird);

    public abstract record InactiveReason
    {
        private InactiveReason() { }

        public sealed record GameOver(int GamePk) : InactiveReason;

        public sealed record DayOff : InactiveReason;

        public sealed record Substituted(int GamePk) : InactiveReason;
    }
}
